### PR TITLE
[ShapeOpt] Feature step size adaption

### PR DESCRIPTION
**Pull request by dbaumgaertner**
This PR:

1. Introduces a simple step size adaption to the steepest descent algorithm. It is based on a comparison of a linear estimation with the actual improvement of the objective. It also utilizes quadratic interpolation.
2. Simplifies a few functions and improves the output of the steepest descent and penalized projection algorithm
3. Adjusts corresponding test to include adaption behavior
4. Adds another test example to test simple algorithms problem setups

### CODE REVIEW

File: applications/ShapeOptimizationApplication/custom_utilities/optimization_utilities.h [hunk 1]
```diff
@@ -79,6 +79,12 @@ class OptimizationUtilities
         // Initialize constraint value
         mConstraintValue = 0.0;
         mPreviousConstraintValue = 0.0;
+
+        // Initialize member variables for penalized projection
+        mCorrectionScaling = 1.0;
```
Comment (philbucher) [inline applications/ShapeOptimizationApplication/custom_utilities/optimization_utilities.h]: why don't you initialize the members when you declare them?

File: applications/ShapeOptimizationApplication/custom_utilities/optimization_utilities.h [hunk 2]
```diff
@@ -79,6 +79,12 @@ class OptimizationUtilities
...
+        std::string algorithm_name = optimizationSettings["optimization_algorithm"]["name"].GetString();
+        if(algorithm_name.compare("penalized_projection") == 0)
```
Comment (philbucher) [inline applications/ShapeOptimizationApplication/custom_utilities/optimization_utilities.h]:
```suggestion
        if(algorithm_name == "penalized_projection")
```

File: applications/ShapeOptimizationApplication/custom_utilities/optimization_utilities.h [hunk 3]
```diff
@@ -99,28 +105,21 @@ class OptimizationUtilities
     // ==============================================================================
     // General optimization operations
     // ==============================================================================
-    void ComputeControlPointUpdate()
+    void ComputeControlPointUpdate(double step_size)
```
Comment (philbucher) [inline applications/ShapeOptimizationApplication/custom_utilities/optimization_utilities.h]:
```suggestion
    void ComputeControlPointUpdate(const double StepSize)
```

File: applications/ShapeOptimizationApplication/custom_utilities/optimization_utilities.h [hunk 4]
```diff
@@ -99,28 +105,21 @@ class OptimizationUtilities
...
-        double step_size = mOptimizationSettings["optimization_algorithm"]["line_search"]["step_size"].GetDouble();
-        bool normalize_search_direction = mOptimizationSettings["optimization_algorithm"]["line_search"]["normalize_search_direction"].GetBool();
-
-        // Computation of update of design variable. Normalization is applied if specified.
-        if(normalize_search_direction)
+        // Normalize if specified
+        if(mOptimizationSettings["optimization_algorithm"]["line_search"]["normalize_search_direction"].GetBool())
         {
-            // Compute max norm of search direction
             double max_norm_search_dir = ComputeMaxNormOfNodalVariable(SEARCH_DIRECTION);
```
Comment (philbucher) [inline applications/ShapeOptimizationApplication/custom_utilities/optimization_utilities.h]:
```suggestion
            const double max_norm_search_dir = ComputeMaxNormOfNodalVariable(SEARCH_DIRECTION);
```

File: applications/ShapeOptimizationApplication/custom_utilities/optimization_utilities.h [hunk 5]
```diff
@@ -99,28 +105,21 @@ class OptimizationUtilities
...
 
-            // Normalize by max norm
             if(max_norm_search_dir>1e-10)
-            {
                 for (auto & node_i : mrDesignSurface.Nodes())
                 {
-                    array_3d normalized_search_direction = node_i.FastGetSolutionStepValue(SEARCH_DIRECTION)/max_norm_search_dir;
-                    noalias(node_i.FastGetSolutionStepValue(SEARCH_DIRECTION)) = normalized_search_direction;
+                    array_3d& search_dir = node_i.FastGetSolutionStepValue(SEARCH_DIRECTION);
```
Comment (philbucher) [inline applications/ShapeOptimizationApplication/custom_utilities/optimization_utilities.h:115]: does SEARCH_DIRECTION have to be historical?

File: applications/ShapeOptimizationApplication/custom_utilities/optimization_utilities.h [hunk 6]
```diff
@@ -292,42 +291,36 @@ class OptimizationUtilities
...
         if(mOptimizationSettings["optimization_algorithm"]["use_adaptive_correction"].GetBool())
         {
-            correction_scaling = AdaptCorrectionScaling( correction_scaling );
-            mOptimizationSettings["optimization_algorithm"]["correction_scaling"].SetDouble(correction_scaling);
+            // Adapt constraint scaling
+
+            // Three cases need to be covered
+            // 1) In case we have two subsequently decreasing constraint values --> correction is fine --> leave current correction scaling
+            // 2) In case the correction jumps over the constraint (change of sign) --> correction was too big --> reduce
+            if(mConstraintValue*mPreviousConstraintValue<0)
```
Comment (philbucher) [inline applications/ShapeOptimizationApplication/custom_utilities/optimization_utilities.h]:
```suggestion
            if (mConstraintValue*mPreviousConstraintValue < 0.0) {
```

File: applications/ShapeOptimizationApplication/custom_utilities/optimization_utilities.h [hunk 7]
```diff
@@ -292,42 +291,36 @@ class OptimizationUtilities
...
+            {
+                mCorrectionScaling *= 0.5;
+                std::cout << "Correction scaling needs to decrease...." << std::endl;
```
Comment (philbucher) [inline applications/ShapeOptimizationApplication/custom_utilities/optimization_utilities.h:300]:
```suggestion
                KRATOS_INFO("some label") << "Correction scaling needs to decrease...." << std::endl;
```

File: applications/ShapeOptimizationApplication/custom_utilities/optimization_utilities.h [hunk 8]
```diff
@@ -292,42 +291,36 @@ class OptimizationUtilities
...
+            }
+            // 3) In case we have subsequently increasing constraint value --> correction was too low --> increase
+            if(std::abs(mConstraintValue)>std::abs(mPreviousConstraintValue) && mConstraintValue*mPreviousConstraintValue>0)
```
Comment (philbucher) [inline applications/ShapeOptimizationApplication/custom_utilities/optimization_utilities.h:303]:
```suggestion
            if(std::abs(mConstraintValue)>std::abs(mPreviousConstraintValue) && mConstraintValue*mPreviousConstraintValue > 0.0)
```

File: applications/ShapeOptimizationApplication/custom_utilities/optimization_utilities.h [hunk 9]
```diff
@@ -292,42 +291,36 @@ class OptimizationUtilities
...
+            {
+                std::cout << "Correction scaling needs to increase...." << std::endl;
```
Comment (philbucher) [inline applications/ShapeOptimizationApplication/custom_utilities/optimization_utilities.h:305]: KRATOS_INFO

File: applications/ShapeOptimizationApplication/custom_utilities/optimization_utilities.h [hunk 10]
```diff
@@ -292,42 +291,36 @@ class OptimizationUtilities
...
+                mCorrectionScaling = std::min(mCorrectionScaling*2,1.0);
+            }
         }
-        return correction_scaling;
+
+        KRATOS_WATCH(mCorrectionScaling)
```
Comment (philbucher) [inline applications/ShapeOptimizationApplication/custom_utilities/optimization_utilities.h:310]: leftover debugging

File: applications/ShapeOptimizationApplication/custom_utilities/optimization_utilities.h [hunk 11]
```diff
@@ -292,42 +291,36 @@ class OptimizationUtilities
...
+
+    	return mCorrectionScaling * norm_search_direction / norm_correction_term;
     }
 
     // --------------------------------------------------------------------------
-    double AdaptCorrectionScaling( double correction_scaling )
+    double GetCorrectionScaling()
```
Comment (philbucher) [inline applications/ShapeOptimizationApplication/custom_utilities/optimization_utilities.h]:
```suggestion
    double GetCorrectionScaling() const
```

File: applications/ShapeOptimizationApplication/python_scripts/algorithm_steepest_descent.py [hunk 12]
```diff
@@ -142,34 +157,83 @@ def __analyzeShape(self):
...
             if self.optimization_iteration == self.max_iterations:
                 print("\n> Maximal iterations of optimization problem reached!")
                 return True
 
+            # Check gradient norm
+            if self.optimization_iteration == 2:
+                self.initial_norm_obj_gradient = self.norm_obj_gradient
+            else:
+                if self.norm_obj_gradient < self.gradient_tolerance*self.initial_norm_obj_gradient:
+                    print("\n> Optimization problem converged as gradient norm reached specified tolerance of ",self.gradient_tolerance)
```
Comment (philbucher) [inline applications/ShapeOptimizationApplication/python_scripts/algorithm_steepest_descent.py:235]:
```suggestion
                    KratosMultiphysics.Logger.PrintInfo("label")("\n> Optimization problem converged as gradient norm reached specified tolerance of ",self.gradient_tolerance)
```

File: applications/ShapeOptimizationApplication/test_examples/08_optimization_plate_with_line_search/analysis_parameters.json [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+      "linear_solver_settings"             : {
+          "solver_type" : "eigen_pardiso_llt",
+          "scaling"     : false,
+          "verbosity"   : 0
+      },
+      "problem_domain_sub_model_part_list" : ["structure"],
+      "processes_sub_model_part_list"      : ["corner_points","load_point"],
+      "rotation_dofs"                      : true
+  },
+  "constraints_process_list" : [{
```
Comment (philbucher) [inline applications/ShapeOptimizationApplication/test_examples/08_optimization_plate_with_line_search/analysis_parameters.json]:
```suggestion
  "processes" : {
     "constraints_process_list" : [{
```

File: applications/ShapeOptimizationApplication/test_examples/08_optimization_plate_with_line_search/analysis_parameters.json [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+          "body_output"         : true,
+          "node_output"         : false,
+          "skin_output"         : false,
+          "plane_output"        : [],
+          "nodal_results"       : ["DISPLACEMENT","REACTION","ROTATION","POINT_LOAD","MESH_CHANGE"],
+          "gauss_point_results" : []
+      },
+      "point_data_configuration"  : []
+  },
+  "restart_options"          : {
```
Comment (philbucher) [inline applications/ShapeOptimizationApplication/test_examples/08_optimization_plate_with_line_search/analysis_parameters.json]: not needed

File: applications/ShapeOptimizationApplication/test_examples/08_optimization_plate_with_line_search/analysis_parameters.json [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+      "SaveRestart"      : false,
+      "RestartFrequency" : 0,
+      "LoadRestart"      : false,
+      "Restart_Step"     : 0
+  },
+  "constraints_data"         : {
```
Comment (philbucher) [inline applications/ShapeOptimizationApplication/test_examples/08_optimization_plate_with_line_search/analysis_parameters.json]: not needed

File: applications/ShapeOptimizationApplication/tests/algorithm_steepest_descent_test/parameters.json [hunk 16]
```diff
@@ -56,19 +56,20 @@
         "optimization_algorithm" : {
             "name"               : "steepest_descent",
             "max_iterations"     : 20,
-            "relative_tolerance" : 1e-0,
+            "relative_tolerance" : 1e-2,
+            "gradient_tolerance" : 0.8,
             "line_search" : {
-                "line_search_type"           : "manual_stepping",
+                "line_search_type"           : "adaptive_stepping",
```
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/tests/algorithm_steepest_descent_test/parameters.json]: can you create a new test for the 'adaptive_stepping' and leave this one as is?

File: applications/ShapeOptimizationApplication/tests/algorithm_penalized_projection_test/parameters.json [hunk 17]
```diff
@@ -62,14 +62,14 @@
         },
         "optimization_algorithm" : {
             "name"                    : "penalized_projection",
-            "correction_scaling"      : 1.0,
+            "correction_scaling"      : 0.2,
```
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/tests/algorithm_penalized_projection_test/parameters.json:65]: Why did you change this? It is not related to the adaptive step_size right?

File: applications/ShapeOptimizationApplication/custom_utilities/optimization_utilities.h (hunk 1 above)
Comment (dbaumgaertner) [reply applications/ShapeOptimizationApplication/custom_utilities/optimization_utilities.h]: Good question. I changed it.

File: applications/ShapeOptimizationApplication/custom_utilities/optimization_utilities.h (hunk 5 above)
Comment (dbaumgaertner) [reply applications/ShapeOptimizationApplication/custom_utilities/optimization_utilities.h:115]: Yes. As it changes every optimization iteration.

File: applications/ShapeOptimizationApplication/python_scripts/algorithm_steepest_descent.py (hunk 12 above)
Comment (dbaumgaertner) [reply applications/ShapeOptimizationApplication/python_scripts/algorithm_steepest_descent.py:235]: I do not see the advantage of using the Logger for such a simple print statement. Why would you want to change that?

File: applications/ShapeOptimizationApplication/custom_utilities/optimization_utilities.h (hunk 5 above)
Comment (philbucher) [reply applications/ShapeOptimizationApplication/custom_utilities/optimization_utilities.h:115]: but do you access the old values at some point, i.e. the ones from a previous iteration?

File: applications/ShapeOptimizationApplication/python_scripts/algorithm_steepest_descent.py (hunk 12 above)
Comment (philbucher) [reply applications/ShapeOptimizationApplication/python_scripts/algorithm_steepest_descent.py:235]: ideally all the prints in Kratos should use the logger
The big advantage is that then you can easily filter, redirect, format and suppress messages in a unified way

if you don't use the logger it will just always print, independent if you want the prints or not
e.g. in a test you don't care abt them, you just want to know if the test passes or not

File: applications/ShapeOptimizationApplication/tests/algorithm_penalized_projection_test/parameters.json (hunk 17 above)
Comment (dbaumgaertner) [reply applications/ShapeOptimizationApplication/tests/algorithm_penalized_projection_test/parameters.json:65]: No, it's not. But I figured out, that in the penalized projection test, the scaling never actually was was adjusted, as the maximum scale of 1 is specified from the very beginning. So I reduced it a bit, to also include the constraint scaling into the testing process.

File: applications/ShapeOptimizationApplication/tests/algorithm_steepest_descent_test/parameters.json (hunk 16 above)
Comment (dbaumgaertner) [reply applications/ShapeOptimizationApplication/tests/algorithm_steepest_descent_test/parameters.json]: Done.

**Comment by dbaumgaertner** [comment]
@armingeiser @philbucher Could you have another look. I corrected some things according to your suggestions. Thanks.

File: applications/ShapeOptimizationApplication/python_scripts/algorithm_steepest_descent.py [hunk 18]
```diff
@@ -142,34 +157,83 @@ def __analyzeShape(self):
...
+    # --------------------------------------------------------------------------
+    def __adjustStepSize(self):
+        current_a = self.step_size
+
+        # Compare actual and estimated improvement using linear information
+        dfda1 = 0.0
+        for node in self.design_surface.Nodes:
+            s1 = node.GetSolutionStepValue(SEARCH_DIRECTION)
+            dfds1 = node.GetSolutionStepValue(DF1DX_MAPPED)
+            dfda1 = dfda1 + s1[0]*dfds1[0] + s1[1]*dfds1[1] + s1[2]*dfds1[2]
```
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/algorithm_steepest_descent.py]: +=

File: applications/ShapeOptimizationApplication/python_scripts/algorithm_steepest_descent.py [hunk 19]
```diff
@@ -142,34 +157,83 @@ def __analyzeShape(self):
...
 
+    # --------------------------------------------------------------------------
+    def __adjustStepSize(self):
+        current_a = self.step_size
+
+        # Compare actual and estimated improvement using linear information
+        dfda1 = 0.0
+        for node in self.design_surface.Nodes:
+            s1 = node.GetSolutionStepValue(SEARCH_DIRECTION)
+            dfds1 = node.GetSolutionStepValue(DF1DX_MAPPED)
```
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/algorithm_steepest_descent.py:169]: maybe add a comment that makes clear that `DF1DX_MAPPED` still has the values from the previous step.

File: applications/ShapeOptimizationApplication/python_scripts/algorithm_steepest_descent.py [hunk 20]
```diff
@@ -142,34 +157,83 @@ def __analyzeShape(self):
...
+
+        df_actual = f2 - f1
+        df_estimated = current_a*dfda1
+
+        # Adjust step size if necessary
+        if f2 < f1:
+            estimation_error = (df_actual-df_estimated)/df_actual
+
+            # Increase step size if linear approximation matches the actual improvement within a specified tolerance
+            if estimation_error < self.approximation_tolerance:
```
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/algorithm_steepest_descent.py]: use consistent naming, `estimation` vs `approximation`

File: applications/ShapeOptimizationApplication/python_scripts/algorithm_steepest_descent.py [hunk 21]
```diff
@@ -142,34 +157,83 @@ def __analyzeShape(self):
...
+            # Leave step size unchanged if a nonliner change in the objective is observed but still a descent direction is obtained
+            else:
+                new_a = current_a
+        else:
+            # Search approximation of optimal step using interpolation
+            a = current_a
+            corrected_step_size = - 0.5 * dfda1 * a**2 / (f2 - f1 - dfda1 * a )
+
+            # Starting from the new design, and assuming an opposite gradient direction, the step size to the approximated optimum behaves reciprocal
+            new_a = current_a-corrected_step_size
```
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/algorithm_steepest_descent.py:195]: :+1:

File: applications/ShapeOptimizationApplication/python_scripts/value_logger_penalized_projection.py [hunk 22]
```diff
@@ -38,7 +38,6 @@ def InitializeLogging( self ):
                 row.append("{:>13s}".format("c"+str(itr+1)+"_ref"))
 
             row.append("{:>13s}".format("c_scaling"))
-            row.append("{:>13s}".format("step_size"))
```
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/value_logger_penalized_projection.py]: i would keep the step size information, even if it is not changing for now

**Review by armingeiser — COMMENTED** [review]
Just a few minor comments.

**Comment by philbucher** [comment]
I don't have further comments

There are still some of my comments I recommend to change, but up to you

**Comment by dbaumgaertner** [comment]
@philbucher Indeed I was not introducing the KratosInfo. This has to be done in a separate PR as it makes sense to do this once for all.

**Comment by dbaumgaertner** [comment]
@armingeiser Another updated according to your comments.

File: applications/ShapeOptimizationApplication/custom_utilities/optimization_utilities.h (hunk 6 above)
Comment (philbucher) [reply applications/ShapeOptimizationApplication/custom_utilities/optimization_utilities.h]: I meant those things

Comment (dbaumgaertner) [reply applications/ShapeOptimizationApplication/custom_utilities/optimization_utilities.h]: Indeed I forgot this. Fixed with the next commit.

**Review by armingeiser — APPROVED** [review]
